extensions/amp-bind/amp-bind.md

```diff
@@ -66,425 +72,972 @@ The `amp-bind` component allows you to add custom stateful interactivity to your
+#### [`Object`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)
+
+Static functions are not namespaced, e.g. use `keys(Object)` instead of `Object.abs(Object)`
+
+- [keys](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys)
+- [values](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values)
+
+[example preview="inline" playground="true" imports="amp-bind"]
 
 ```html
-<amp-state id="myState">
+<amp-state id="myObjectState">
   <script type="application/json">
     {
+      "hello": "world",
       "foo": "bar"
     }
   </script>
 </amp-state>
+<p [text]="'keys: ' + keys(myObjectState)">
+  keys: myObjectState JSON object keys
+</p>
+<p [text]="'values: ' + values(myObjectState)">
+  values: myObjectState JSON object values
+</p>
+<button on="tap:AMP.setState({})">Evaluate</button>
+```
+
+[/example]
+
+#### [`Global`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects)
+
+- [encodeURI](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI)
+- [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent)
+
+[example preview="inline" playground="true" imports="amp-bind"]
+
+```html
+<p [text]="'encodeURI: ' + encodeURI('https://amp.dev/😉')">
+  encodeURI: Encode a URI and ignore protocol prefix
+</p>
+<p [text]="'encodeURIComponent: ' + encodeURIComponent('https://amp.dev/😉')">
+  encodeURIComponent: Encode a URI
+</p>
+<button on="tap:AMP.setState({})">Evaluate</button>
+```
+
+[/example]
+
+### Defining macros with `amp-bind-macro`
+
+Reuse `amp-bind` expression fragments by defining an `amp-bind-macro`. The `amp-bind-macro` element allows an expression that takes zero or more arguments and references the current state. Invoke `amp-bind-macros` like a function, referencing the `id` attribute value from anywhere in the document.
+
+[example preview="inline" playground="true" imports="amp-bind"]
+
+```html
+<amp-bind-macro
+  id="circleArea"
+  arguments="radius"
+  expression="3.14 * radius * radius"
+></amp-bind-macro>
+<p>
+  Input a radius value
+</p>
+<input
+  type="number"
+  min="0"
+  max="100"
+  value="0"
+  on="input-throttled:AMP.setState({myCircle:{radius: event.value}})"
+/>
+<p>
+  The circle has an area of
+  <span [text]="circleArea(myCircle.radius)">0</span>.
+</p>
 ```
 
-[Expressions](#expressions) can reference state variables via dot syntax. In this example, `myState.foo` will evaluate to `"bar"`.
+[/example]
 
-- An `<amp-state>` element's child JSON has a maximum size of 100KB.
-- An `<amp-state>` element can also specify a CORS URL instead of a child JSON script. See the [Appendix](#amp-state-specification) for details.
+A macro can also call other macros <i>defined before itself</i>. A macro cannot call itself recursively.
```

Comment:
Is there a limit to the number of macros a macro can call? Worth documenting in a followup?

Reply:
also added to notes